File: src/state.rs
use sqlx::PgPool;

use crate::config::AppConfig;
use crate::services::cache::CacheService;
use crate::services::progress_buffer::ProgressBuffer;
use crate::services::session_manager::SessionManager;
use crate::services::scan_progress::ScanProgressTracker;
use crate::services::probe_task::ProbeTask;
use crate::services::file_watcher::FileWatcher;
use crate::services::log_buffer::LogBuffer;
use crate::services::tmdb::ScrapeTask;

pub struct AppState {
    pub db: PgPool,
    pub cache: CacheService,
    pub config: AppConfig,
    pub session_manager: SessionManager,
    pub progress_buffer: ProgressBuffer,
    pub scan_progress: ScanProgressTracker,
    pub probe_task: ProbeTask,
    pub file_watcher: FileWatcher,
    pub log_buffer: std::sync::Arc<LogBuffer>,
    pub scrape_task: ScrapeTask,
    pub http_client: reqwest::Client,
}

File: internal/handlers/state.go
package handlers

import (
	"net/http"

	"github.com/gin-gonic/gin"
	"github.com/jackc/pgx/v5/pgxpool"

	"fyms/internal/config"
	"fyms/internal/services"
	"fyms/internal/services/sysmetrics"
	"fyms/internal/services/taskcenter"
)

type AppState struct {
	DB             *pgxpool.Pool
	Cache          *services.CacheService
	Config         *config.AppConfig
	SessionManager *services.SessionManager
	ProgressBuffer *services.ProgressBuffer
	ScanProgress   *services.ScanProgressTracker
	ProbeTask      *services.ProbeTask
	FileWatcher    *services.FileWatcher
	Ingest         *services.IngestWorker
	ScrapeQueue    *services.ScrapeQueue
	ScrapeWorker   *services.ScrapeWorker
	LogBuffer      *services.LogBuffer
	ScrapeTask     *services.ScrapeTask
	HTTPClient     *http.Client
	Updater        *services.Updater
	GapScanTask    *services.GapScanTask
	BackfillTask   *services.BackfillTask
	TaskCenter     *taskcenter.Registry
	TaskChain      *taskcenter.ChainEngine
	SysMetrics     *sysmetrics.Collector
}

func GetState(c *gin.Context) *AppState {
	v, _ := c.Get("state")
	return v.(*AppState)
}
